Allow user workload monitoring configuration ConfigMap to be created in UWM ns

lilic opened the pull request:
* [X] I added CHANGELOG entry for this change.

lilic commented:
/retest

s-urbaniak commented on cmd/operator/main.go:
```diff
@@ -89,6 +89,7 @@ func Main() int {
 	namespaceUserWorkload := flagset.String("namespace-user-workload", "openshift-user-workload-monitoring", "Namespace to deploy and manage user workload monitoring stack in.")
 	namespaceSelector := flagset.String("namespace-selector", "openshift.io/cluster-monitoring=true", "Selector for namespaces to monitor.")
 	configMapName := flagset.String("configmap", "cluster-monitoring-config", "ConfigMap name to configure the cluster monitoring stack.")
+	userWorkloadConfigMapName := flagset.String("userWorkloadConfigmap", "user-workload-monitoring-config", "ConfigMap name to configure the user workload monitoring stack.")
```
why does this need to be configurable?

lilic replied:
Just followed pattern we had, see line above here, honestly happy to leave it out as well!

s-urbaniak replied:
my intuition is we should limit exposure here as we want to move to a CRD based setup :thinking: 

This cli flag increases the public API surfaces.

lilic replied:
good point, removed it!

s-urbaniak commented on pkg/manifests/config.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 The Cluster Monitoring Operator Authors
+/// Copyright 2018 The Cluster Monitoring Operator Authors
```
The triple slash is unneeded i believe.

lilic replied:
I have no idea what happened here :D lets see if its my weird vim configuration, nice catch

s-urbaniak commented on pkg/operator/operator.go:
```diff
@@ -355,6 +360,32 @@ func (o *Operator) sync(key string) error {
+			return manifests.NewDefaultUserWorkloadMonitoringConfig(), nil
+		} else {
+			klog.Warningf("Error loading User Workload Monitoring ConfigMap. Error: %v", err)
+			return nil, err
```
let's wrap the returning error to have context.

s-urbaniak commented on pkg/operator/operator.go:
```diff
@@ -355,6 +360,32 @@ func (o *Operator) sync(key string) error {
+		uwc, err := manifests.NewUserConfigFromString(configContent)
+		if err != nil {
+			klog.Warningf("Error creating User Workload Configuration from ConfigMap. Error: %v", err)
+			return nil, err
```
dito

s-urbaniak commented on pkg/operator/operator.go:
```diff
@@ -387,8 +419,19 @@ func (o *Operator) Config(key string) (*manifests.Config, error) {
+	if *c.ClusterMonitoringConfiguration.UserWorkloadEnabled {
+		c.UserWorkloadConfiguration, err = o.loadUserWorkloadConfig()
+		if err != nil {
+
```
i believe this is WIP

s-urbaniak replied:
or isn't it? :-) do we really want to swallow the error?

lilic replied:
nope, just me forgetting to return error :( fixed

s-urbaniak commented:
looking great so far! :tada:

s-urbaniak commented:
/lgtm

lilic commented:
@s-urbaniak PTAL changes since last time:
- I removed the flag, good call!
- added example configmap and used it in one of the config unit tests
- added CHANGELOG entry.

Thanks!

lilic commented:
/hold cancel

lilic commented:
/retest

simonpasquier commented on examples/user-workload/1q:
```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
```
I'm confused about this file's name. Is it needed?

lilic replied:
I believe that was committed wrongly, possibly outcome of some generation or just my mistake, will remove.

simonpasquier commented on pkg/manifests/config.go:
```diff
@@ -31,23 +31,28 @@ type Config struct {
-	PrometheusK8sConfig          *PrometheusK8sConfig `json:"prometheusK8s"`
-	PrometheusUserWorkloadConfig *PrometheusK8sConfig `json:"prometheusUserWorkload"`
+	ClusterMonitoringConfiguration *ClusterMonitoringConfiguration `json:"-"`
+	UserWorkloadConfiguration      *UserWorkloadConfiguration      `json:"-"`
```
Having both `UserWorkloadConfiguration` and `UserWorkloadConfig` is confusing.

lilic replied:
UserWorkloadConfig?

simonpasquier replied:
We have a `UserWorkloadConfig` struct:
https://github.com/openshift/cluster-monitoring-operator/blob/882dc1b7310edd7981fab2aff76fdc30672a470c/pkg/manifests/config.go#L157-L159

And another `UserWorkloadConfiguration` struct:
https://github.com/openshift/cluster-monitoring-operator/blob/882dc1b7310edd7981fab2aff76fdc30672a470c/pkg/manifests/config.go#L381-L385

The names are close which makes it confusing for me when reviewing the code.

lilic replied:
I see, makes sense! Any suggestions for new or old name?

simonpasquier commented on pkg/manifests/config.go:
```diff
@@ -31,23 +31,28 @@ type Config struct {
+	ThanosQuerierConfig      *ThanosQuerierConfig         `json:"thanosQuerier"`
+	UserWorkloadEnabled      *bool                        `json:"enableUserWorkload"`
+	// TODO: Remove in 4.7 release.
+	PrometheusUserWorkloadConfig         *PrometheusK8sConfig      `json:"prometheusUserWorkload"`
```
Instead of duplicating `PrometheusUserWorkloadConfig`, `PrometheusOperatorUserWorkloadConfig` and `ThanosRulerConfig` here and in `UserWorkloadConfiguration` maybe it would be simpler to use only `UserWorkloadConfiguration`? IOW handle the "legacy" fields only when loading the config map but don't expose them here.

lilic replied:
The point was to be able to easily remove the current logic, hence the TODO comments, as we will stop supporting this in 4.7. And to separate out the two clear tenants we have right now cluster and user workload.

simonpasquier replied:
An alternative approach would be to remove the user workload monitoring fields from the `ClusterMonitoringConfiguration` struct and only use `Config.UserWorkloadConfiguration`.  `NewConfig()` would try to load the uwm fields until we decide that it shouldn't.

Something like this: https://github.com/simonpasquier/cluster-monitoring-operator/commit/44bd3ddbed3f2ee07ef99d239e15f1b599ccb581

The side effect is that uwm configuration will be taken either from the "legacy" fields or from the new configmap in `openshift-user-workload-monitoring` but it won't be a merge of both. In the current state, `PrometheusOperatorConfig` can be defined in the `openshift-montoring` configmap and `PrometheusConfig` in the `openshift-user-workload-monitoring` configmap, it can be confusing or desired, depending on how you see things :)

(IOW: in other words)

brancz replied:
hmm .. especially with configuration I much prefer duplication structs. This is a lesson learned the hard way in Kubernetes. Only because things happen to be identical at this point in time, doesn't mean they may not very well develop in different direction in the future, and they almost always do unless the concept is literally the same (eg. TLS configuration in Prometheus). In this case we even know it will be deleted in the future, so I actually prefer @lilic 's currently proposed way.

simonpasquier commented on pkg/manifests/manifests.go:
```diff
@@ -1404,39 +1408,77 @@ func (f *Factory) PrometheusUserWorkload(grpcTLS *v1.Secret) (*monv1.Prometheus,
+
+	// TODO: remove after 4.7
+
+	if f.config.ClusterMonitoringConfiguration.PrometheusUserWorkloadConfig.LogLevel != "" {
```
When you have config fields defined both in `f.config.ClusterMonitoringConfiguration`  and `f.config.UserWorkloadConfiguration` then the former will take precedence. Couldn't this cause confusion?

lilic replied:
techPreview is what we have to support right now primarily hence why it takes precedence here, but that is a good point to document.

lilic commented:
/retest 
🤔

lilic commented on pkg/operator/operator.go:
```diff
@@ -355,6 +360,32 @@ func (o *Operator) sync(key string) error {
+	userCM, err := o.client.GetConfigmap(o.namespaceUserWorkload, o.userWorkloadConfigMapName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Warning("No User Workload Monitoring ConfigMap was found. Using defaults.")
```
We use warning level for the cluster monitoring configmap, I just followed that pattern. It makes sense to me in a way, as you are warning user they did not configure their stack at all. But don't have too strong opinion so can change, but then we should do both?

simonpasquier replied:
I'd rather have the message at the info level because for me a warning is something I need at least to investigate and not customizing the monitoring isn't something suspicious. At info, the message would still logged so it would be available for support requests.
And +1000 that if we change it here, it should be consistent across the board.
